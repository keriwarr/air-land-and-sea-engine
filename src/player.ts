import { enumValues } from 'utils';

export enum PLAYER {
  ONE = 'ONE',
  TWO = 'TWO',
}

export const PLAYERS = enumValues(PLAYER);

export const getOtherPlayer = (player: PLAYER) =>
  player === PLAYER.ONE ? PLAYER.TWO : PLAYER.ONE;
