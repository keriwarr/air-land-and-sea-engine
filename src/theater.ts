import { enumValues, exhaustiveSwitch } from './utils';

export enum THEATER {
  AIR = 'AIR',
  LAND = 'LAND',
  SEA = 'SEA',
}
export const THEATERS = enumValues(THEATER);

export const getDifferentTheater = (theater: THEATER) => {
  switch (theater) {
    case THEATER.AIR:
      return THEATER.LAND;
    case THEATER.LAND:
      return THEATER.SEA;
    case THEATER.SEA:
      return THEATER.AIR;
    default:
      return exhaustiveSwitch({
        switchValue: theater,
      });
  }
};
