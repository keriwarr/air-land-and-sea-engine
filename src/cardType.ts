import { enumValues, mapToObject } from './utils';
import { THEATER } from './theater';

export enum CARD_EFFECT_TYPE {
  INSTANT = 'INSTANT',
  ONGOING = 'ONGOING',
  REGULAR = 'REGULAR',
}

export enum CARD_TYPE_KEY {
  SUPPORT = 'SUPPORT',
  AIR_DROP = 'AIR_DROP',
  MANEUVER = 'MANEUVER',
  AERODROME = 'AERODROME',
  CONTAINMENT = 'CONTAINMENT',
  HEAVY = 'HEAVY',
  REINFORCE = 'REINFORCE',
  AMBUSH = 'AMBUSH',
  COVER_FIRE = 'COVER_FIRE',
  DISRUPT = 'DISRUPT',
  TRANSPORT = 'TRANSPORT',
  ESCALATION = 'ESCALATION',
  REDEPLOY = 'REDEPLOY',
  BLOCKADE = 'BLOCKADE',
}
const CARD_TYPE_KEYS = enumValues(CARD_TYPE_KEY);

interface ICardType {
  theater?: THEATER;
  rank: number;
  defaultName: string;
  effectType: CARD_EFFECT_TYPE;
  description: string | null;
}

const CARD_TYPE: { readonly [K in CARD_TYPE_KEY]: ICardType } = {
  [CARD_TYPE_KEY.SUPPORT]: {
    theater: THEATER.AIR,
    rank: 1,
    defaultName: 'Support',
    effectType: CARD_EFFECT_TYPE.ONGOING,
    description: 'You gain +3 strength in each adjacent theater.',
  },
  [CARD_TYPE_KEY.AIR_DROP]: {
    theater: THEATER.AIR,
    rank: 2,
    defaultName: 'Air Drop',
    effectType: CARD_EFFECT_TYPE.INSTANT,
    description:
      'On your next turn, you may play a battle card to a non-matching theater.',
  },
  [CARD_TYPE_KEY.MANEUVER]: {
    rank: 3,
    defaultName: 'Maneuver',
    effectType: CARD_EFFECT_TYPE.INSTANT,
    description: 'FLIP a Battle Card in an adjacent theater.',
  },
  [CARD_TYPE_KEY.AERODROME]: {
    theater: THEATER.AIR,
    rank: 4,
    defaultName: 'Aerodrome',
    effectType: CARD_EFFECT_TYPE.ONGOING,
    description:
      'You may play Battle Cards of strength 3 or less to non-matching theaters.',
  },
  [CARD_TYPE_KEY.CONTAINMENT]: {
    theater: THEATER.AIR,
    rank: 5,
    defaultName: 'Containment',
    effectType: CARD_EFFECT_TYPE.ONGOING,
    description:
      'If either player players a Battle Card face-down, immediately discard that card',
  },
  [CARD_TYPE_KEY.HEAVY]: {
    rank: 6,
    defaultName: 'Heavy',
    effectType: CARD_EFFECT_TYPE.REGULAR,
    description: null,
  },
  [CARD_TYPE_KEY.REINFORCE]: {
    theater: THEATER.LAND,
    rank: 1,
    defaultName: 'Reinforce',
    effectType: CARD_EFFECT_TYPE.INSTANT,
    description:
      'Look at the top card of the Battle Deck. You may play it face down to an adjacent theater.',
  },
  [CARD_TYPE_KEY.AMBUSH]: {
    theater: THEATER.LAND,
    rank: 2,
    defaultName: 'Ambush',
    effectType: CARD_EFFECT_TYPE.INSTANT,
    description: 'FLIP a Battle Card in any theater.',
  },
  [CARD_TYPE_KEY.COVER_FIRE]: {
    theater: THEATER.LAND,
    rank: 4,
    defaultName: 'Cover Fire',
    effectType: CARD_EFFECT_TYPE.ONGOING,
    description: 'All Battle Cards covered by this card are now strength 4.',
  },
  [CARD_TYPE_KEY.DISRUPT]: {
    theater: THEATER.LAND,
    rank: 5,
    defaultName: 'Disrupt',
    effectType: CARD_EFFECT_TYPE.INSTANT,
    description:
      'Your opponent chooses and FLIPS 1 of their Battle Cards and then you FLIP 1 of yours.',
  },
  [CARD_TYPE_KEY.TRANSPORT]: {
    theater: THEATER.SEA,
    rank: 1,
    defaultName: 'Transport',
    effectType: CARD_EFFECT_TYPE.INSTANT,
    description: 'You may MOVE 1 of your Battle Cards to a different theater.',
  },
  [CARD_TYPE_KEY.ESCALATION]: {
    theater: THEATER.SEA,
    rank: 2,
    defaultName: 'Escalation',
    effectType: CARD_EFFECT_TYPE.ONGOING,
    description: 'All of your face-down cards are now Strength 4.',
  },
  [CARD_TYPE_KEY.REDEPLOY]: {
    theater: THEATER.SEA,
    rank: 4,
    defaultName: 'Redeploy',
    effectType: CARD_EFFECT_TYPE.INSTANT,
    description:
      'Return 1 of your face-down cards to your hand. If you do, gain an extra turn.',
  },
  [CARD_TYPE_KEY.BLOCKADE]: {
    theater: THEATER.SEA,
    rank: 5,
    defaultName: 'Blockade',
    effectType: CARD_EFFECT_TYPE.ONGOING,
    description:
      "If a Battle Card is played in an adjacent theater with 3 or more cards already in it (counting both players' cards), discard that card with no effect.",
  },
};

export class CardType implements ICardType {
  readonly key: CARD_TYPE_KEY;
  readonly theater: ICardType['theater'];
  readonly rank: ICardType['rank'];
  readonly defaultName: ICardType['defaultName'];
  readonly effectType: ICardType['effectType'];
  readonly description: ICardType['description'];

  private constructor(key: CARD_TYPE_KEY) {
    this.key = key;

    const cardType = CARD_TYPE[key];
    if (cardType.theater) {
      this.theater = cardType.theater;
    }
    this.rank = cardType.rank;
    this.defaultName = cardType.defaultName;
    this.effectType = cardType.effectType;
    this.description = cardType.description;
  }

  static readonly byKey = mapToObject(CARD_TYPE_KEYS, key =>
    Object.freeze(new CardType(key))
  );
  static readonly allKeys = CARD_TYPE_KEYS;
  static readonly all = CardType.allKeys.map(key => CardType.byKey[key]);
}
