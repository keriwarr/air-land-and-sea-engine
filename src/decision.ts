import { PLAYER, getOtherPlayer } from './player';
import { THEATER } from './theater';
import { CARD_TYPE_KEY } from './cardType';
import { IBoardState } from './board';
import { enumValues } from './utils';

export enum DECISION_TYPE {
  FLIP_DECISION = 'FLIP_DECISION',
  REINFORCE_DECISION = 'REINFORCE_DECISION',
  TRANSPORT_DECISION = 'TRANSPORT_DECISION',
  REDEPLOY_DECISION = 'REDEPLOY_DECISION',
}

interface IGenericDecision<T extends DECISION_TYPE = DECISION_TYPE> {
  type: T;
}

export interface IFlipDecision
  extends IGenericDecision<DECISION_TYPE.FLIP_DECISION> {
  targetedPlayer: PLAYER;
  theater: THEATER;
}
export interface IReinforceDecision
  extends IGenericDecision<DECISION_TYPE.REINFORCE_DECISION> {
  made: {
    theater: THEATER;
  } | null;
}
export interface ITransportDecision
  extends IGenericDecision<DECISION_TYPE.TRANSPORT_DECISION> {
  made: {
    originTheater: THEATER;
    originIndexFromTop: number;
    destinationTheater: THEATER;
  } | null;
}
export interface IRedeployDecision
  extends IGenericDecision<DECISION_TYPE.REDEPLOY_DECISION> {
  made: {
    theater: THEATER;
    indexFromTop: number;
  };
}

export type IDecision =
  | IFlipDecision
  | IReinforceDecision
  | ITransportDecision
  | IRedeployDecision;

export interface IAnticipatedDecision {
  type: DECISION_TYPE;
  player: PLAYER;
  promptingMoveIndex: number;
}

export const getAnticipatedDecisions = (
  cardTypeKey: CARD_TYPE_KEY,
  player: PLAYER,
  promptingMoveIndex: number,
  playedTheater: THEATER,
  boardState: IBoardState,
  getAdjacentTheaters: (theater: THEATER) => THEATER[]
): IAnticipatedDecision[] => {
  const anticipatedTypeAndPlayers = (() => {
    switch (cardTypeKey) {
      case CARD_TYPE_KEY.REINFORCE:
        return [{ type: DECISION_TYPE.REINFORCE_DECISION, player }];
      case CARD_TYPE_KEY.AMBUSH:
        return [{ type: DECISION_TYPE.FLIP_DECISION, player }];
      case CARD_TYPE_KEY.MANEUVER: {
        const adjacentTheaters = getAdjacentTheaters(playedTheater);
        const adjacentPlayerTheaters = adjacentTheaters
          .map(theater =>
            enumValues(PLAYER).map(player => boardState[theater][player])
          )
          .reduce((flat, playerTheaters) => [...flat, ...playerTheaters], []);
        if (
          !adjacentPlayerTheaters.find(
            adjacentPlayerTheater => adjacentPlayerTheater.length > 0
          )
        ) {
          return [];
        }
        return [{ type: DECISION_TYPE.FLIP_DECISION, player }];
      }
      case CARD_TYPE_KEY.DISRUPT:
        return [
          { type: DECISION_TYPE.FLIP_DECISION, player },
          {
            type: DECISION_TYPE.FLIP_DECISION,
            player: getOtherPlayer(player),
          },
        ];
      case CARD_TYPE_KEY.REDEPLOY: {
        const playerCardStates = enumValues(THEATER)
          .map(theater => boardState[theater][player])
          .reduce((flat, cardStates) => [...flat, ...cardStates], []);
        if (!playerCardStates.find(cardState => !cardState.faceUp)) {
          return [];
        }
        return [{ type: DECISION_TYPE.REDEPLOY_DECISION, player }];
      }
      case CARD_TYPE_KEY.TRANSPORT:
        return [{ type: DECISION_TYPE.TRANSPORT_DECISION, player }];
      default:
        return [];
    }
  })();

  return anticipatedTypeAndPlayers.map(typeAndPlayer => ({
    ...typeAndPlayer,
    promptingMoveIndex: promptingMoveIndex,
  }));
};
