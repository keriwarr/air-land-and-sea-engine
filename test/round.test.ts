import { RoundState } from '../src/round';
import { PLAYER } from '../src/player';
import { DECISION_TYPE } from '../src/decision';
import { THEATER } from '../src/theater';
import { CARD_TYPE_KEY } from '../src/cardType';
import { autorun } from 'mobx';

describe('RoundState', () => {
  describe('End Conditions', () => {
    let roundState: RoundState;

    beforeEach(() => {
      roundState = new RoundState([THEATER.AIR, THEATER.LAND, THEATER.SEA]);
    });

    it('initially has no victor', () => {
      expect(roundState.victor).toBe(null);
      expect(roundState.complete).toBe(false);
    });

    it('has a victor after a surrender', () => {
      roundState.surrender();
      expect(roundState.victor).toBe(PLAYER.TWO);
      expect(roundState.complete).toBe(true);
    });

    it('has a different victor after a later surrender', () => {
      autorun(reaction => {
        roundState.playCard(roundState.currentHandP1[0].getMove());

        reaction.dispose();
      });
      roundState.surrender();
      expect(roundState.victor).toBe(PLAYER.ONE);
      expect(roundState.complete).toBe(true);
    });

    it('does not permit surrendering twice', () => {
      roundState.surrender();
      expect(() => {
        roundState.surrender();
      }).toThrow();
    });
  });

  describe('Cards Types', () => {
    let roundState: RoundState;

    beforeEach(() => {
      roundState = new RoundState([THEATER.AIR, THEATER.LAND, THEATER.SEA], {
        disableHandContainsCheck: true,
      });
    });

    describe(CARD_TYPE_KEY.SUPPORT, () => {
      it('adds strength to the center theater', () => {
        roundState.playCard(
          roundState.deck
            .find({ type: CARD_TYPE_KEY.SUPPORT })
            .getMove({ theater: THEATER.AIR })
        );

        expect(roundState.orderedTheaterStrengths).toMatchInlineSnapshot(`
          Array [
            Object {
              "ONE": 1,
              "TWO": 0,
              "theater": "AIR",
            },
            Object {
              "ONE": 3,
              "TWO": 0,
              "theater": "LAND",
            },
            Object {
              "ONE": 0,
              "TWO": 0,
              "theater": "SEA",
            },
          ]
        `);
      });

      it('adds strength to the outside theaters', () => {
        roundState = new RoundState([THEATER.LAND, THEATER.AIR, THEATER.SEA], {
          disableHandContainsCheck: true,
        });
        roundState.playCard(
          roundState.deck
            .find({ type: CARD_TYPE_KEY.SUPPORT })
            .getMove({ theater: THEATER.AIR })
        );

        expect(roundState.orderedTheaterStrengths).toMatchInlineSnapshot(`
          Array [
            Object {
              "ONE": 3,
              "TWO": 0,
              "theater": "LAND",
            },
            Object {
              "ONE": 1,
              "TWO": 0,
              "theater": "AIR",
            },
            Object {
              "ONE": 3,
              "TWO": 0,
              "theater": "SEA",
            },
          ]
        `);
      });

      it('does not work when face down', () => {
        roundState.playCard(
          roundState.deck
            .find({ type: CARD_TYPE_KEY.SUPPORT })
            .getMove({ theater: THEATER.AIR, faceUp: false })
        );

        expect(roundState.orderedTheaterStrengths).toMatchInlineSnapshot(`
          Array [
            Object {
              "ONE": 2,
              "TWO": 0,
              "theater": "AIR",
            },
            Object {
              "ONE": 0,
              "TWO": 0,
              "theater": "LAND",
            },
            Object {
              "ONE": 0,
              "TWO": 0,
              "theater": "SEA",
            },
          ]
        `);
      });

      it('stops working when flipped over', () => {
        roundState.playCard(
          roundState.deck
            .find({ type: CARD_TYPE_KEY.SUPPORT })
            .getMove({ theater: THEATER.AIR })
        );
        roundState.playCard(
          roundState.deck.find({ type: CARD_TYPE_KEY.AMBUSH }).getMove()
        );
        roundState.playDecision({
          decision: {
            type: DECISION_TYPE.FLIP_DECISION,
            targetedPlayer: PLAYER.ONE,
            theater: THEATER.AIR,
          },
        });

        expect(roundState.orderedTheaterStrengths).toMatchInlineSnapshot(`
          Array [
            Object {
              "ONE": 2,
              "TWO": 0,
              "theater": "AIR",
            },
            Object {
              "ONE": 0,
              "TWO": 2,
              "theater": "LAND",
            },
            Object {
              "ONE": 0,
              "TWO": 0,
              "theater": "SEA",
            },
          ]
        `);

        roundState.playCard(
          roundState.deck
            .find({
              type: CARD_TYPE_KEY.MANEUVER,
              theater: THEATER.LAND,
            })
            .getMove()
        );
        roundState.playDecision({
          decision: {
            type: DECISION_TYPE.FLIP_DECISION,
            targetedPlayer: PLAYER.ONE,
            theater: THEATER.AIR,
          },
        });

        expect(roundState.orderedTheaterStrengths).toMatchInlineSnapshot(`
          Array [
            Object {
              "ONE": 1,
              "TWO": 0,
              "theater": "AIR",
            },
            Object {
              "ONE": 6,
              "TWO": 2,
              "theater": "LAND",
            },
            Object {
              "ONE": 0,
              "TWO": 0,
              "theater": "SEA",
            },
          ]
        `);
      });
    });

    describe(CARD_TYPE_KEY.AIR_DROP, () => {
      it('permits playing in to non-matching theaters', () => {
        roundState.playCard(
          roundState.deck.find({ type: CARD_TYPE_KEY.AIR_DROP }).getMove()
        );

        roundState.playCard(
          roundState.deck
            .find({ type: CARD_TYPE_KEY.HEAVY, theater: THEATER.LAND })
            .getMove()
        );

        expect(() => {
          roundState.playCard(
            roundState.deck
              .find({ type: CARD_TYPE_KEY.HEAVY, theater: THEATER.AIR })
              .getMove({ theater: THEATER.LAND }),
            { dryRun: true }
          );
        }).not.toThrow();

        expect(() => {
          roundState.playCard(
            roundState.deck
              .find({ type: CARD_TYPE_KEY.HEAVY, theater: THEATER.AIR })
              .getMove({ theater: THEATER.SEA }),
            { dryRun: true }
          );
        }).not.toThrow();

        expect(() => {
          roundState.playCard(
            roundState.deck
              .find({ type: CARD_TYPE_KEY.HEAVY, theater: THEATER.SEA })
              .getMove({ theater: THEATER.AIR }),
            { dryRun: true }
          );
        }).not.toThrow();
      });

      it.todo("doesn't override blockade");

      it.todo("doesn't override containment");

      it('can be cancelled by flipping', () => {
        roundState.playCard(
          roundState.deck
            .find({ type: CARD_TYPE_KEY.AIR_DROP })
            .getMove({ theater: THEATER.AIR })
        );

        roundState.playCard(
          roundState.deck
            .find({ type: CARD_TYPE_KEY.AMBUSH, theater: THEATER.LAND })
            .getMove()
        );

        roundState.playDecision({
          decision: {
            type: DECISION_TYPE.FLIP_DECISION,
            targetedPlayer: PLAYER.ONE,
            theater: THEATER.AIR,
          },
        });

        expect(() => {
          roundState.playCard(
            roundState.deck
              .find({ type: CARD_TYPE_KEY.HEAVY, theater: THEATER.AIR })
              .getMove({ theater: THEATER.SEA }),
            { dryRun: true }
          );
        }).toThrow();

        expect(() => {
          roundState.playCard(
            roundState.deck
              .find({ type: CARD_TYPE_KEY.HEAVY, theater: THEATER.LAND })
              .getMove({ theater: THEATER.AIR }),
            { dryRun: true }
          );
        }).toThrow();

        expect(() => {
          roundState.playCard(
            roundState.deck
              .find({ type: CARD_TYPE_KEY.HEAVY, theater: THEATER.SEA })
              .getMove({ theater: THEATER.LAND }),
            { dryRun: true }
          );
        }).toThrow();
      });

      it("doesn't prevent playing in matching theaters", () => {
        roundState.playCard(
          roundState.deck.find({ type: CARD_TYPE_KEY.AIR_DROP }).getMove()
        );

        roundState.playCard(
          roundState.deck.find({ type: CARD_TYPE_KEY.COVER_FIRE }).getMove()
        );

        expect(() => {
          roundState.playCard(
            roundState.deck
              .find({ type: CARD_TYPE_KEY.HEAVY, theater: THEATER.AIR })
              .getMove({ theater: THEATER.AIR }),
            { dryRun: true }
          );
        }).not.toThrow();

        expect(() => {
          roundState.playCard(
            roundState.deck
              .find({ type: CARD_TYPE_KEY.HEAVY, theater: THEATER.LAND })
              .getMove({ theater: THEATER.LAND }),
            { dryRun: true }
          );
        }).not.toThrow();

        expect(() => {
          roundState.playCard(
            roundState.deck
              .find({ type: CARD_TYPE_KEY.HEAVY, theater: THEATER.SEA })
              .getMove({ theater: THEATER.SEA }),
            { dryRun: true }
          );
        }).not.toThrow();
      });
    });

    describe(CARD_TYPE_KEY.MANEUVER, () => {
      it("can't flip over itself", () => {
        roundState.playCard(
          roundState.deck
            .find({ type: CARD_TYPE_KEY.MANEUVER, theater: THEATER.AIR })
            .getMove()
        );

        expect(() => {
          roundState.playDecision({
            decision: {
              type: DECISION_TYPE.FLIP_DECISION,
              targetedPlayer: PLAYER.ONE,
              theater: THEATER.AIR,
            },
          });
        }).toThrow();
      });

      it.todo("can't flip over a card in a non-adjacent theater");

      it.todo('can filp over an allied card');

      it.todo('can filp over an enemy card');

      it.todo('can be flipped over by a triggered effect');

      it.todo('has no effect if there are no targetable cards');

      it.todo('must flip a card if there are targetable cards');
    });

    describe(CARD_TYPE_KEY.AERODROME, () => {
      it.todo('permits cards to be played in any theater');

      it.todo("doesn't grant permissions to the opponent");

      it.todo("doesn't work on cards of strength four or more");

      it.todo('stops working when flipped over');
    });

    describe(CARD_TYPE_KEY.CONTAINMENT, () => {
      it.todo('causes a relevant status effect to be advertised');

      it.todo("doesn't prevent cards from being played face-down");

      it.todo('discards cards that are played face down');

      it.todo('prevents discarded cards from using their effect');

      it.todo(
        'prevents a decision from being anticipated when causing a discard'
      );

      it.todo('can be disabled by being flipped over');

      it.todo("doesn't cause flipped cards to be discarded");

      it.todo('stops working when flipped over');
    });

    describe(CARD_TYPE_KEY.HEAVY, () => {
      it("just adds strength where it's played", () => {
        roundState.playCard(
          roundState.deck
            .find({ type: CARD_TYPE_KEY.HEAVY, theater: THEATER.AIR })
            .getMove({ theater: THEATER.AIR })
        );

        expect(roundState.orderedTheaterStrengths).toMatchInlineSnapshot(`
          Array [
            Object {
              "ONE": 6,
              "TWO": 0,
              "theater": "AIR",
            },
            Object {
              "ONE": 0,
              "TWO": 0,
              "theater": "LAND",
            },
            Object {
              "ONE": 0,
              "TWO": 0,
              "theater": "SEA",
            },
          ]
        `);
      });
    });

    describe(CARD_TYPE_KEY.REINFORCE, () => {
      it.todo('adds a card face down');

      it.todo('can add a card that is immediately discarded');

      it.todo('allows choosing not to reinforce');

      it.todo('must add the card in an adjacent theater');

      it.todo('can be triggered multiple times');
    });

    describe(CARD_TYPE_KEY.AMBUSH, () => {
      it.todo('can flip over itself');

      it.todo('can flip over a card in an adjacent theater');

      it.todo('can flip over a card in a non-adjacent theater');

      it.todo('can filp over an allied card');

      it.todo('can filp over an enemy card');

      it.todo('can be flipped over by a triggered effect');

      it.todo('must always flip over a card');
    });

    describe(CARD_TYPE_KEY.COVER_FIRE, () => {
      it.todo('strengthens weak cards beneath it');

      it.todo('strengthens flipped cards beneath it');

      it.todo('weakens strong cards beneath it');

      it.todo('stops working when flipped over');

      it.todo('works as expected when moved to a new theater');
    });

    describe(CARD_TYPE_KEY.DISRUPT, () => {
      it.todo('causes two decisions to be enqueued');

      it.todo('can cause the opponent to flip a card in any theater');

      it.todo("let's the opponent choose which card to flip");

      it.todo('requires the opponent to flip a card if able');

      it.todo('allows the opponent to flip a card in any theater');

      it.todo(
        'resovles the effects of the opponents flip before proceeding to your flip'
      );

      it.todo("then let's you choose which card to flip");

      it.todo('requires you to flip a card');

      it.todo('it allows you to flip a card in any theater');

      it.todo('can flip over a card in a non-adjacent theater');

      it.todo('can be cancelled by being flipped over mid-effect');
    });

    describe(CARD_TYPE_KEY.TRANSPORT, () => {
      it.todo('moves a card');

      it.todo('can move face down cards');

      it.todo('can move covered cards');

      it.todo('does not trigger containment or blockade');

      it.todo('permits choosing not to transport');

      it.todo('can transport itself');

      it.todo('can be triggered multiple times');
    });

    describe(CARD_TYPE_KEY.ESCALATION, () => {
      it.todo('strengthens your flipped down cards');

      it.todo("doesn't strengthen the opponents flipped down cards");

      it.todo("doesn't strengthen face-up cards");

      it.todo('stops working when flipped over');

      it.todo('works as expected when moved to a new theater');
    });

    describe(CARD_TYPE_KEY.REDEPLOY, () => {
      it.todo('returns a card to your hand');

      it.todo("can't return a face-up card to your hand");

      it.todo('must return a card if possible');

      it.todo('does not grant a bonus turn if a card is not returned');

      it.todo('grants an immediate bonus turn when played');

      it.todo('grants an immedaite bonus turn when flipped by its owner');

      it.todo('grants a delayed bonus turn when flipped by its opponent');

      it.todo('can be activated more than once');

      it.todo('handles a truly degenerate case');
    });

    describe(CARD_TYPE_KEY.BLOCKADE, () => {
      it.todo('causes a relevant status effect to be advertised');

      it.todo("doesn't prevent cards from being played in affected theaters");

      it.todo('discards cards that are played in affected theaters');

      it.todo('prevents discarded cards from using their effect');

      it.todo(
        'prevents a decision from being anticipated when causing a discard'
      );

      it.todo('can be disabled by being flipped over');

      it.todo('affects both players');

      it.todo('stops working when flipped over');

      it.todo('works as expected when moved to a new theater');
    });
  });

  describe('Move Validation', () => {
    let roundState: RoundState;

    beforeEach(() => {
      roundState = new RoundState([THEATER.AIR, THEATER.LAND, THEATER.SEA]);
    });

    it("prevents playing cards that are not in the player's hand", () => {
      autorun(reaction => {
        expect(() => {
          roundState.playCard(roundState.currentHandP2[0].getMove());
        }).toThrow();

        reaction.dispose();
      });

      // TODO - I'd like to test this case too, but I'd need a way to
      // conveniently play whatever decisions are necessary after playing p1's
      // first card.
      // roundState.playCard(roundState.currentHandP1[0].getMove());
      // expect(() => {
      //   roundState.playCard(roundState.currentHandP1[0].getMove());
      // }).toThrow();
    });

    it.todo('prevents playing cards face up to non-matching theaters');

    it.todo("doesn't prevent playing cards that would be discarded");

    it.todo('prevents playing cards when a decision is anticipated');

    it.todo('prevents surrendering when a decision is anticipated');

    it.todo('prevents make a decision when none is anticipated');

    it.todo('prevents making a decision of the incorrect type');

    it.todo(
      'prevents making a decision of the correct type but from the incorrect player'
    );

    it.todo('prevents playing cards once a player has surrendered');
  });

  describe('Discarding', () => {
    it.todo('moves the discarded card to a visible pile');
  });

  describe('Face-down Cards', () => {
    it.todo('can be played in any theater');

    it.todo('has strength of two');

    it.todo("doesn't trigger instant effects");

    it.todo('allows instant effects to be triggered when flipped up');

    it.todo('keeps track of whether the opponent has seen which card it is');
  });

  describe('Hand', () => {
    it.todo('loses a card when it is played');
  });

  describe('Deck', () => {
    it.todo("knows which cards haven't been seen by either player");

    it.todo('start with six cards in it');

    it.todo('loses a card each time a reinforce decision is made');
  });

  describe('Testing Features', () => {
    describe('playMove:dryRun', () => {});

    describe('opts.disableHandContainsCheck', () => {});
  });

  describe('Status Effects', () => {
    describe('Global Status Effects', () => {
      it.todo('has a status effect when Containment is played');
    });

    describe('Player Status Effects', () => {
      it.todo('has a status effect when Air Drop is played');

      it.todo('has a status effect when Escalation is played');

      it.todo('has a status effect when Aerodrome is played');
    });

    describe('Theater Status Effects', () => {
      it.todo('has a status effect when Blockade is played');

      it.todo('has a status effect when Support is played');
    });
  });
});
