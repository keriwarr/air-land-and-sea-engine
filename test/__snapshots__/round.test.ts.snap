// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoundState Import/Export Exports a complete history of the round to JSON 1`] = `
Object {
  "deck": Array [
    Object {
      "cardTypeKey": "BLOCKADE",
      "description": "If a Battle Card is played in an adjacent theater with 3 or more cards already in it (counting both players' cards), discard that card with no effect.",
      "effectType": "ONGOING",
      "id": 17,
      "name": "Blockade",
      "rank": 5,
      "theater": "SEA",
    },
    Object {
      "cardTypeKey": "HEAVY",
      "description": null,
      "effectType": "REGULAR",
      "id": 12,
      "name": "Heavy Tanks",
      "rank": 6,
      "theater": "LAND",
    },
    Object {
      "cardTypeKey": "COVER_FIRE",
      "description": "All Battle Cards covered by this card are now strength 4.",
      "effectType": "ONGOING",
      "id": 10,
      "name": "Cover Fire",
      "rank": 4,
      "theater": "LAND",
    },
    Object {
      "cardTypeKey": "AMBUSH",
      "description": "FLIP a Battle Card in any theater.",
      "effectType": "INSTANT",
      "id": 8,
      "name": "Ambush",
      "rank": 2,
      "theater": "LAND",
    },
    Object {
      "cardTypeKey": "HEAVY",
      "description": null,
      "effectType": "REGULAR",
      "id": 18,
      "name": "Super Battleship",
      "rank": 6,
      "theater": "SEA",
    },
    Object {
      "cardTypeKey": "HEAVY",
      "description": null,
      "effectType": "REGULAR",
      "id": 6,
      "name": "Heavy Bombers",
      "rank": 6,
      "theater": "AIR",
    },
    Object {
      "cardTypeKey": "AIR_DROP",
      "description": "On your next turn, you may play a battle card to a non-matching theater.",
      "effectType": "INSTANT",
      "id": 2,
      "name": "Air Drop",
      "rank": 2,
      "theater": "AIR",
    },
    Object {
      "cardTypeKey": "MANEUVER",
      "description": "FLIP a Battle Card in an adjacent theater.",
      "effectType": "INSTANT",
      "id": 3,
      "name": "Maneuver",
      "rank": 3,
      "theater": "AIR",
    },
    Object {
      "cardTypeKey": "TRANSPORT",
      "description": "You may MOVE 1 of your Battle Cards to a different theater.",
      "effectType": "INSTANT",
      "id": 13,
      "name": "Transport",
      "rank": 1,
      "theater": "SEA",
    },
    Object {
      "cardTypeKey": "MANEUVER",
      "description": "FLIP a Battle Card in an adjacent theater.",
      "effectType": "INSTANT",
      "id": 9,
      "name": "Maneuver",
      "rank": 3,
      "theater": "LAND",
    },
    Object {
      "cardTypeKey": "AERODROME",
      "description": "You may play Battle Cards of strength 3 or less to non-matching theaters.",
      "effectType": "ONGOING",
      "id": 4,
      "name": "Aerodrome",
      "rank": 4,
      "theater": "AIR",
    },
    Object {
      "cardTypeKey": "CONTAINMENT",
      "description": "If either player players a Battle Card face-down, immediately discard that card.",
      "effectType": "ONGOING",
      "id": 5,
      "name": "Containment",
      "rank": 5,
      "theater": "AIR",
    },
    Object {
      "cardTypeKey": "MANEUVER",
      "description": "FLIP a Battle Card in an adjacent theater.",
      "effectType": "INSTANT",
      "id": 15,
      "name": "Maneuver",
      "rank": 3,
      "theater": "SEA",
    },
    Object {
      "cardTypeKey": "ESCALATION",
      "description": "All of your face-down cards are now Strength 4.",
      "effectType": "ONGOING",
      "id": 14,
      "name": "Escalation",
      "rank": 2,
      "theater": "SEA",
    },
    Object {
      "cardTypeKey": "DISRUPT",
      "description": "Your opponent chooses and FLIPS 1 of their Battle Cards and then you FLIP 1 of yours.",
      "effectType": "INSTANT",
      "id": 11,
      "name": "Disrupt",
      "rank": 5,
      "theater": "LAND",
    },
    Object {
      "cardTypeKey": "REDEPLOY",
      "description": "Return 1 of your face-down cards to your hand. If you do, gain an extra turn.",
      "effectType": "INSTANT",
      "id": 16,
      "name": "Redeploy",
      "rank": 4,
      "theater": "SEA",
    },
    Object {
      "cardTypeKey": "SUPPORT",
      "description": "You gain +3 strength in each adjacent theater.",
      "effectType": "ONGOING",
      "id": 1,
      "name": "Support",
      "rank": 1,
      "theater": "AIR",
    },
    Object {
      "cardTypeKey": "REINFORCE",
      "description": "Look at the top card of the Battle Deck. You may play it face down to an adjacent theater.",
      "effectType": "INSTANT",
      "id": 7,
      "name": "Reinforce",
      "rank": 1,
      "theater": "LAND",
    },
  ],
  "moves": Array [
    Object {
      "faceUp": true,
      "id": 17,
      "theater": "SEA",
      "type": "CARD",
    },
    Object {
      "faceUp": true,
      "id": 12,
      "theater": "LAND",
      "type": "CARD",
    },
    Object {
      "faceUp": true,
      "id": 10,
      "theater": "LAND",
      "type": "CARD",
    },
    Object {
      "faceUp": false,
      "id": 8,
      "theater": "LAND",
      "type": "CARD",
    },
    Object {
      "faceUp": true,
      "id": 18,
      "theater": "SEA",
      "type": "CARD",
    },
    Object {
      "faceUp": true,
      "id": 6,
      "theater": "AIR",
      "type": "CARD",
    },
    Object {
      "faceUp": true,
      "id": 2,
      "theater": "AIR",
      "type": "CARD",
    },
    Object {
      "faceUp": false,
      "id": 3,
      "theater": "AIR",
      "type": "CARD",
    },
    Object {
      "faceUp": true,
      "id": 13,
      "theater": "SEA",
      "type": "CARD",
    },
    Object {
      "decision": Object {
        "made": Object {
          "destinationTheater": "LAND",
          "originIndexFromTop": 0,
          "originTheater": "AIR",
        },
        "type": "TRANSPORT_DECISION",
      },
      "type": "DECISION",
    },
  ],
}
`;
